cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/internal/replication/PullStrategyMockTest.java

```diff
@@ -231,6 +232,97 @@ public Object answer(InvocationOnMock invocation) throws Throwable {
     }
 
+    @Test
+    public void testSetCheckpointWhenEmpty() throws Exception {
+        CouchDB mockRemoteDb = mock(CouchDB.class);
+        when(mockRemoteDb.changes((PullFilter) null, null, 1000)).then(new Answer<Object>() {
+            @Override
+            public Object answer(InvocationOnMock invocation) throws Throwable {
+                FileReader fr = new FileReader(TestUtils.loadFixture
+                        ("fixture/empty_changes.json"));
+                return JSONUtils.fromJson(fr, ChangesResult.class);
+            }
+        });
+        when(mockRemoteDb.exists()).thenReturn(true);
+
+        StrategyListener mockListener = mock(StrategyListener.class);
+        PullStrategy pullStrategy = super.getPullStrategy();
+        pullStrategy.sourceDb = mockRemoteDb;
+        pullStrategy.getEventBus().register(mockListener);
+        pullStrategy.run();
+
+        //should have 0 document
+        Assert.assertEquals(this.datastore.getDocumentCount(), 0);
+        //Checkpoint should be created in targetDb
+        String checkpoint =
+                (String) pullStrategy.targetDb.getCheckpoint(pullStrategy.getReplicationId());
+        Assert.assertEquals(checkpoint, "10-d9e5b0147af143e5b6d1979378ad957b");
+        //make sure the correct events were fired
+        verify(mockListener).complete(any(ReplicationStrategyCompleted.class));
+        verify(mockListener, never()).error(any(ReplicationStrategyErrored.class));
+    }
+
+    @Test
+    public void testDoNotSetCheckpointWhenNotModified() throws Exception {
+        CouchDB mockRemoteDb = mock(CouchDB.class);
+        when(mockRemoteDb.changes((PullFilter) null, "10-d9e5b0147af143e5b6d1979378ad957b", 1000)).then(new Answer<Object>() {
+            @Override
+            public Object answer(InvocationOnMock invocation) throws Throwable {
+                FileReader fr = new FileReader(TestUtils.loadFixture
+                        ("fixture/empty_changes.json"));
+                return JSONUtils.fromJson(fr, ChangesResult.class);
+            }
+        });
+        when(mockRemoteDb.exists()).thenReturn(true);
+
+        DatastoreWrapper mockLocalDb = mock(DatastoreWrapper.class);
+        when(mockLocalDb.getCheckpoint(any(String.class))).thenReturn("10" +
+                "-d9e5b0147af143e5b6d1979378ad957b");
+
+        StrategyListener mockListener = mock(StrategyListener.class);
+        PullStrategy pullStrategy = super.getPullStrategy();
+        pullStrategy.sourceDb = mockRemoteDb;
+        pullStrategy.targetDb = mockLocalDb;
+        pullStrategy.getEventBus().register(mockListener);
+        pullStrategy.run();
+
+        //make sure the correct events were fired
+        verify(mockListener).complete(any(ReplicationStrategyCompleted.class));
+        verify(mockListener, never()).error(any(ReplicationStrategyErrored.class));
+        verify(mockLocalDb, never()).putCheckpoint(anyString(), any());
```

Comment:
This test failed on device and from some investigations it seems that mocking of a package-private class doesn't work with mockito in Dalvik because of some classloader differences (see e.g. https://github.com/mockito/mockito/issues/796 and https://github.com/janstk/dexmaker/issues/26#issuecomment-142492955).
Given that this is a mock test anyway I don't think we're gaining anything by running it on device since it runs elsewhere anyway, so I suggest just using an assume to skip the test if mocking fails e.g.:
```java
        try {
            DatastoreWrapper mockLocalDb = mock(DatastoreWrapper.class);
            when(mockLocalDb.getCheckpoint(any(String.class))).thenReturn("10" +
                    "-d9e5b0147af143e5b6d1979378ad957b");

            StrategyListener mockListener = mock(StrategyListener.class);
            PullStrategy pullStrategy = super.getPullStrategy();
            pullStrategy.sourceDb = mockRemoteDb;
            pullStrategy.targetDb = mockLocalDb;
            pullStrategy.getEventBus().register(mockListener);
            pullStrategy.run();

            //make sure the correct events were fired
            verify(mockListener).complete(any(ReplicationStrategyCompleted.class));
            verify(mockListener, never()).error(any(ReplicationStrategyErrored.class));
            verify(mockLocalDb, never()).putCheckpoint(anyString(), any());
        } catch(UnsupportedOperationException uoe) {
            assumeNoException("Cannot proxy DatastoreWrapper on Dalvik", uoe);
        }
```

Reply:
Hi @ricellis, I've included your propoal and tested on the emulator, now both tests are skipped.
```
com.cloudant.sync.internal.replication.PullStrategyMockTest > testSetCheckpointWhenModified[Nexus_4_API_23(AVD) - 6.0] SKIPPED 
  org.junit.internal.AssumptionViolatedException: Cannot proxy DatastoreWrapper on Dalvik: got: <java.lang.UnsupportedOperationException: can
not proxy inaccessible class class com.cloudant.sync.internal.replication.DatastoreWrapper>, expected: null
        at org.junit.Assume.assumeThat(Assume.java:118)

com.cloudant.sync.internal.replication.PullStrategyMockTest > testSetCheckpointWhenModified[Nexus_4_API_23(AVD) - 6.0] SKIPPED 
    org.junit.internal.AssumptionViolatedException: Cannot proxy DatastoreWrapper on Dalvik: got: <java.lang.UnsupportedOperationException: cann
ot proxy inaccessible class class com.cloudant.sync.internal.replication.DatastoreWrapper>, expected: null
        at org.junit.Assume.assumeThat(Assume.java:118)
```

Thanks